hoptimator-jdbc/src/main/java/com/linkedin/hoptimator/jdbc/HoptimatorConnection.java

```diff
@@ -68,4 +93,24 @@ private void registerMaterialization(List<String> viewPath, RelNode tableRel, Re
   public List<RelOptMaterialization> materializations() {
     return materializations;
   }
+
+  /** A logger that logs to both SLF4J and an internal list. */
+  public static class HoptimatorConnectionDualLogger {
+    private final Logger slf4jLogger;
+
+    public HoptimatorConnectionDualLogger(Class<?> clazz) {
+      this.slf4jLogger = org.slf4j.LoggerFactory.getLogger(clazz);
+    }
+
+    private final List<String> logs = new ArrayList<>();
+
+    public void info(String format, Object... arguments) {
+      logs.add(String.format(format, arguments));
```

Comment:
See above wrt hooks. Instead of `logs.add()` here, you could just invoke the hook: `infoLogHook.accept(msg)`.